File: samples/java/src/main/java/co/windly/limbosample/presentation/base/fragment/base/BaseFragment.java
package co.windly.limbosample.presentation.base.fragment.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import butterknife.ButterKnife;
import co.windly.limbo.fragment.base.LimboFragment;
import co.windly.limbo.fragment.base.LimboFragmentView;
import co.windly.limbo.presenter.queue.LimboQueuePresenter;
import dagger.Lazy;
import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasAndroidInjector;
import dagger.android.support.AndroidSupportInjection;
import javax.inject.Inject;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@SuppressWarnings("WeakerAccess")
public abstract class BaseFragment<V extends LimboFragmentView, P extends LimboQueuePresenter<V>>
  extends LimboFragment<V, P> implements LimboFragmentView, HasAndroidInjector {

  //region Android Injector

  @Inject
  protected DispatchingAndroidInjector<Object> androidInjector;

  @Override
  public AndroidInjector<Object> androidInjector() {
    return androidInjector;
  }

  //endregion

  //region Presenter

  @Inject
  @SuppressWarnings("WeakerAccess")
  protected Lazy<P> fragmentPresenter;

  @NonNull
  @Override
  public P createPresenter() {
    return fragmentPresenter.get();
  }

  //endregion

  //region Lifecycle

  @Override
  public void onAttach(@NotNull Context context) {

    // Inject dependencies.
    AndroidSupportInjection.inject(this);

    super.onAttach(context);
  }

  @Nullable
  @Override
  public View onCreateView(@NotNull LayoutInflater inflater, @Nullable ViewGroup container,
    @Nullable Bundle savedInstanceState) {

    // Create view.
    final View view = super.onCreateView(inflater, container, savedInstanceState);

    // Throw an exception if view cannot be inflated.
    if (view == null) {
      throw new IllegalStateException("Unable to inflate view.");
    }

    // Bind views.
    ButterKnife.bind(this, view);

    // Return created view.
    return view;
  }

  //endregion
}

File: samples/kotlin/src/main/java/co/windly/limbosample/presentation/base/fragment/base/BaseFragment.kt
package co.windly.limbosample.presentation.base.fragment.base

import android.content.Context
import co.windly.limbo.mvp.fragment.base.LimboFragment
import co.windly.limbo.mvp.fragment.base.LimboFragmentView
import co.windly.limbo.mvp.presenter.queue.LimboQueuePresenter
import dagger.Lazy
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import dagger.android.support.AndroidSupportInjection
import javax.inject.Inject

abstract class BaseFragment<V : LimboFragmentView, P : LimboQueuePresenter<V>> : LimboFragment<V, P>(),
  LimboFragmentView, HasAndroidInjector {

  //region Android Injector

  @Inject
  lateinit var androidInjector: DispatchingAndroidInjector<Any>

  override fun androidInjector(): AndroidInjector<Any> =
    androidInjector

  //endregion

  //region Presenter

  @Inject
  lateinit var fragmentPresenter: Lazy<P>

  override fun createPresenter(): P =
    fragmentPresenter.get()

  //endregion

  //region Lifecycle

  override fun onAttach(context: Context) {

    // Inject dependencies.
    AndroidSupportInjection.inject(this)

    super.onAttach(context)
  }

  //endregion
}
